In [2]:
import sklearn

In [1]:
from sklearn.datasets import load_iris

In [2]:
iris_dataset = load_iris() # 샘플데이터 저장
iris_dataset

{'data': array([[5.1, 3.5, 1.4, 0.2],
        [4.9, 3. , 1.4, 0.2],
        [4.7, 3.2, 1.3, 0.2],
        [4.6, 3.1, 1.5, 0.2],
        [5. , 3.6, 1.4, 0.2],
        [5.4, 3.9, 1.7, 0.4],
        [4.6, 3.4, 1.4, 0.3],
        [5. , 3.4, 1.5, 0.2],
        [4.4, 2.9, 1.4, 0.2],
        [4.9, 3.1, 1.5, 0.1],
        [5.4, 3.7, 1.5, 0.2],
        [4.8, 3.4, 1.6, 0.2],
        [4.8, 3. , 1.4, 0.1],
        [4.3, 3. , 1.1, 0.1],
        [5.8, 4. , 1.2, 0.2],
        [5.7, 4.4, 1.5, 0.4],
        [5.4, 3.9, 1.3, 0.4],
        [5.1, 3.5, 1.4, 0.3],
        [5.7, 3.8, 1.7, 0.3],
        [5.1, 3.8, 1.5, 0.3],
        [5.4, 3.4, 1.7, 0.2],
        [5.1, 3.7, 1.5, 0.4],
        [4.6, 3.6, 1. , 0.2],
        [5.1, 3.3, 1.7, 0.5],
        [4.8, 3.4, 1.9, 0.2],
        [5. , 3. , 1.6, 0.2],
        [5. , 3.4, 1.6, 0.4],
        [5.2, 3.5, 1.5, 0.2],
        [5.2, 3.4, 1.4, 0.2],
        [4.7, 3.2, 1.6, 0.2],
        [4.8, 3.1, 1.6, 0.2],
        [5.4, 3.4, 1.5, 0.4],
        [5.2, 4.1, 1.5, 0.1],
  

In [17]:
print(f'iris_dataset key : {iris_dataset.keys()}')
print('iris_dataset key :', iris_dataset.keys())

iris_dataset key : dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename'])
iris_dataset key : dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename'])


In [26]:
# print(iris_dataset['data'])
# print(iris_dataset['data'].shape)
# print(iris_dataset['feature_names']) # 항목에 대한 설명 확인
# print(iris_dataset['target'])
# print(iris_dataset['target_names']) # 0 : setosa / 1 : versicolor / 2 : virginica
print(iris_dataset['DESCR'])

.. _iris_dataset:

Iris plants dataset
--------------------

**Data Set Characteristics:**

    :Number of Instances: 150 (50 in each of three classes)
    :Number of Attributes: 4 numeric, predictive attributes and the class
    :Attribute Information:
        - sepal length in cm
        - sepal width in cm
        - petal length in cm
        - petal width in cm
        - class:
                - Iris-Setosa
                - Iris-Versicolour
                - Iris-Virginica
                
    :Summary Statistics:

    ============== ==== ==== ======= ===== ====================
                    Min  Max   Mean    SD   Class Correlation
    ============== ==== ==== ======= ===== ====================
    sepal length:   4.3  7.9   5.84   0.83    0.7826
    sepal width:    2.0  4.4   3.05   0.43   -0.4194
    petal length:   1.0  6.9   3.76   1.76    0.9490  (high!)
    petal width:    0.1  2.5   1.20   0.76    0.9565  (high!)
    ============== ==== ==== ======= ===== ===========

In [4]:
# model_selection 패키지의 train_test_split을 가져와서 데이터를 쪼갠다. 
from sklearn.model_selection import train_test_split

데이터를 쪼개면 train data -> x, y 와  test data -> x, y 이렇게 4조각으로 나온다.

In [8]:
# 학습용 입력, 테스트용 입력, 학습용 타겟, 테스트용 타겟 변수를 지정
train_input, test_input, train_label, test_label = \
train_test_split(iris_dataset['data'], iris_dataset['target'], test_size=0.25, random_state=42)

x, y, test set의 사이즈 (75%는 train set, 25%는 test set이 된다) 보통 0.2 ~ 0.3 사이<br/>
random_state ? 데이터가 한쪽으로 치우쳐지지 않게 랜덤으로 섞어줘야 한다.<br/>
random_state에 들어가는 숫자는 정해진 게 없다. 얼마나 랜덤하게 돌릴건지 라고 생각하면 된다.

쪼개진 데이터를 shape으로 확인해본다<br/><br/>
train_data<br/>
총 150개 데이터 중 75%인 112개가 학습되고, 25%인 38개가 테스트로 쓰였다.<br/>
feature는 4개인 것을 확인할 수 있다.<br/>
train과 test의 타겟은 0, 1, 2 중 하나의 값을 가지기 때문에 차원은 하나뿐이다.

In [6]:
print(train_input.shape)
print(test_input.shape)
print(train_label.shape)
print(test_label.shape)

(112, 4)
(38, 4)
(112,)
(38,)


# 최근접 알고리즘
학습을 했는데 데이터들이 특정부분마다 뭉쳐있는 것을 확인할 수 있다.<br/>
새로운 데이터를 추가했을 때 어느쪽에 더 가까운지 보고 편입시켜주는 알고리즘이다.<br/>
근처에 있는 것을 몇개까지 기준으로 잡을것인지가 중요하다.

알고리즘을 어떻게 선택했을까?<br/>
<img src="https://scikit-learn.org/stable/_static/ml_map.png" style="float:left">

In [19]:
from sklearn.neighbors import KNeighborsClassifier

# 알고리즘을 메모리에 가져온다.
knn = KNeighborsClassifier(n_neighbors=3)

구별할 때 근처에 있는 것 한개를 기준으로 본다.<br/>
워낙 가벼운 알고리즘이고 빠르게 처리될 것이다.<br/><br/>
knn 학습을 시킬 때 fit이라는 명령을 주면 학습을 한다.<br/>
학습에 필요한 train_input과 train_target 값을 줘야한다.

In [20]:
knn.fit(train_input, train_label)

KNeighborsClassifier(n_neighbors=3)

학습이 끝난 후에 predict를 하면 예측을 하기 시작한다.<br/><br/>
예측치를 predict_label에 받고,<br/>
따로 만들어놓은 test용 데이터를 넣어서 예측을 돌린다.

In [21]:
predict_label = knn.predict(test_input)

In [22]:
print(predict_label) # 텍스트 값을 줬을 때 예측된 결과
print(test_label) # 실제 값

[1 0 2 1 1 0 1 2 1 1 2 0 0 0 0 1 2 1 1 2 0 2 0 2 2 2 2 2 0 0 0 0 1 0 0 2 1
 0]
[1 0 2 1 1 0 1 2 1 1 2 0 0 0 0 1 2 1 1 2 0 2 0 2 2 2 2 2 0 0 0 0 1 0 0 2 1
 0]


예측 값과 실제 값, 두개를 비교했을 때 얼마나 일치하는 지 봐야한다. <br/>
100% 일치하는 게 좋은 건 아니다.<br/>
과적합이 되었을 가능성이 농후하다.<br/>
데이터를 바꿨을 때 오차가 어마어마하게 나올 수 있다.<br/>

# 정확도 계산

In [23]:
import numpy as np

print(f'test accuracy : {np.mean(predict_label == test_label)}')

test accuracy : 1.0


예측값과 테스트 값이 일치하는 지 일치율의 평균을 구해본다.<br/>
accuracy = 1은 정확도 100%라는 것이다.<br/>
iris꽃은 90% 정확도가 넘는 데이터이기 때문에 거의 100% 나올 것이다.

전체 작업의 80%의 작업과 모델을 만드는 것은 20%가 채 안되는 작업이다.<br/>
대부분 80%의 작업은 데이터를 정제하는 작업이다.<br/>